rotkehlchen/chain/ethereum/modules/balancer/decoder.py

```diff
@@ -0,0 +1,291 @@
+        return None, [action_item]
+
+    def decode_balancer_v2_event(
+            self,
+            tx_log: EvmTxReceiptLog,
+            transaction: EvmTransaction,  # pylint: disable=unused-argument
+            decoded_events: list[HistoryBaseEntry],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+            action_items: list[ActionItem],
+    ) -> tuple[Optional[HistoryBaseEntry], list[ActionItem]]:
+        if tx_log.topics[0] == V2_SWAP:
+            return self._decode_swap_creation(
+                tx_log=tx_log,
+                decoded_events=decoded_events,
+                action_items=action_items,
+            )
+        return None, []
+
+    def _maybe_enrich_balancer_v2_transfers(  # pylint: disable=no-self-use
+            self,
+            token: 'EvmToken',  # pylint: disable=unused-argument
+            tx_log: EvmTxReceiptLog,
+            transaction: EvmTransaction,
+            event: HistoryBaseEntry,
+            action_items: list[ActionItem],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+    ) -> bool:
+        """
+        Enrich tranfer transactions to address for jar deposits and withdrawals
+        May raise:
+        - UnknownAsset
+        - WrongAssetType
+        """
+        if action_items is None or len(action_items) == 0 or transaction.to_address != VAULT_ADDRESS:  # noqa: E501
+            return False
+
+        assert action_items[-1].extra_data is not None
+        asset = event.asset.resolve_to_evm_token()
+        to_amount = asset_normalized_value(
+            amount=action_items[-1].extra_data['amount_out'],
+            asset=asset,
+        )
+
+        if (
+            action_items[-1].extra_data['to_token'] != tx_log.address or
+            event.balance.amount != to_amount
+        ):
+            return False
+
+        is_receive = asset.evm_address == action_items[-1].extra_data['to_token']
+
+        event.counterparty = CPT_BALANCER_V2
+        event.event_type = HistoryEventType.TRADE
+        if is_receive:
+            event.event_subtype = HistoryEventSubType.RECEIVE
+            event.notes = f'Receive {event.balance.amount} {asset.symbol} in balancer V2 from {event.location_label}'  # noqa: E501
+        else:
+            event.event_subtype = HistoryEventSubType.SPEND
+
+        return True
+
+    def _maybe_enrich_balancer_v1_events(
+        self,
+        token: 'EvmToken',
+        tx_log: EvmTxReceiptLog,  # pylint: disable=unused-argument
+        transaction: EvmTransaction,  # pylint: disable=unused-argument
+        event: HistoryBaseEntry,
+        action_items: list[ActionItem],  # pylint: disable=unused-argument
+        all_logs: list[EvmTxReceiptLog],
+    ) -> bool:
+        actions = self._decode_v1_pool_event(all_logs=all_logs)
+        if actions is None:
+            return False
+
+        for action in actions:
+            assert action.extra_data is not None
+            if (
```

Comment:
holy shit that shit is fuuuuuuuuuuuuuuuuuuugly.

Is there no other way to make it easier to read? Perhaps a dict?

Reply:
I've tried to simplify a bit the logic, added more comments explaining things and removed the assertions